gribi helpers

mojiiba opened the pull request:
This pull requests adds helpers (initial version) for gribi to simplify writing gribi test cases. Please let me know if you find the helpers useful and you are willing to merge the PR when it is completed.

liulk commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+	t.Logf("closing GRIBI connection for dut: %s", g.dut.Name())
+	g.fluentC.Stop(t)
+	v := reflect.Indirect(reflect.ValueOf(g.gribiC)).FieldByName("cc")
+	clientConn, ok := (reflect.NewAt(v.Type(), unsafe.Pointer(v.UnsafeAddr())).Elem().Interface()).(*grpc.ClientConn)
```
I see what the problem is. The Ondatra API only gives us the client stub, and it is not designed to close the underlying connection if it's new. For now, let's use the Default() stub, but consider filing an issue with Ondatra to rectify the RawAPIs design.

Accessing unexported fields using reflection is problematic.

mojiiba replied:
@liulk  using  Default() is not going to help here since Ondatra does not provide close for default gribi client as well.
I have discussed this with @greg-dennis (here)[https://github.com/openconfig/ondatra/pull/3] and he suggested using reflection. Anyway, If you agree we have this code here until we finalize the solution for this.

greg-dennis replied:
Yeah, I suggested getting the connection field by reflection because the need to close the connection seemed like a rather rare requirement, making it hard to justify the additional weight and complexity on the binding that would be needed to support it. We also have no resources to design the changes to accommodate the use case at this time, and so far this is the one and only request for that feature. I don't love it, but I also don't see the generated proto code changing in a way to break it any time soon.

Two suggestions:
(1) I suspect "reflect.Indirect(reflect.ValueOf(g.gribiC))" would be better as "reflect.ValueOf(g.gribiC).Elem()"
(2) Does every test really need to close the connection? I thought it was a fairly narrow test case that actually needed the connection closed. Might be better to confine this to just that test case rather than making it part of a common library. The above-mentioned gribigo/fluent library has found no need to ever close the connection.

liulk replied:
You can definitely reuse the gRIBI client stub over many fluent clients. The abstraction that defines a gRIBI session (election ID, persistence) is contained entirely within the Modify streaming RPC.

Hence, `RawAPIs().GRIBI().Default()` is the better way. There is no need to close this default transport since it's just in the background and indefinitely reusable.

mojiiba replied:
Thanks for the clarification!  I changed the code as you suggested. There is no need to use reflection.

liulk commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+	return g.lastElectionID.Low, g.lastElectionID.High
+}
+
+func (g *GRIBIHandler) BeMaster(t *testing.T) {
```
Suggested new name: `BecomeLeader`

mojiiba replied:
done

liulk commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+// set the election id
+func (g *GRIBIHandler) UpdateElectionID(t *testing.T, lowElecId, highElecId uint64) {
```
`t *testing.T` should be `t testing.TB` consistently throughout.

mojiiba replied:
done

liulk commented:
Overall I think this helper is going to be really useful. We have a few tests that could be migrated to use this helper also. Here is one that we open sourced recently: https://github.com/openconfig/featureprofiles/tree/main/feature/gribi/ate_tests/base_leader_election_test

mojiiba commented:
@liulk Thank you so much for the comments. Sure, I will address the comment and migrate the published test case  to use the new helpers.

liulk commented:
I just remembered that some of the tests prefer to batch each route's NH, NHG, and IPv4Entry into one request with the modify-await-chk, and these tests might not gain very much from this particular helper.

* Route 1: NH1, NHG1, IPv4E1
* Route 2: NH2, NHG2, IPv4E2
* etc.

In other words, the batching in those tests is done in row order.

But this particular helper batches NH1, NH2, etc. into a batch for modify-await-chk, then NHG1, NHG2, etc. as a batch, then IPv4E1, IPv4E2, etc. as a batch. The batching is done in columnar order.

Use judgment whether a test will benefit from this helper before trying to migrate them.

mojiiba commented:
> 

How about having a separate function (AddRoute) to cover the batch cases as well?

liulk commented:
That could work!

mojiiba commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+	g.UpdateElectionID(t, lowElecId, highElecId)
+}
+
+func (g *GRIBIHandler) AddNHG(t *testing.T, nhgIndex uint64, nhs map[uint64]uint64, instance string) {
```
done

mojiiba commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+	return g.fluentC.Await(subctx, t)
+}
+
+// get the election id (last known election id) from a dut
```
done

mojiiba commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+func (g *GRIBIHandler) awaitTimeout(ctx context.Context, t testing.TB, timeout time.Duration) error {
```
done

mojiiba commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+package gribi
```
done

greg-dennis replied:
How is this all different from
https://github.com/openconfig/gribigo/blob/main/fluent/fluent.go
?

I'm not sure we should have two fluent gribi APIs in openconfig.

liulk replied:
I think the value proposition of this package is that it streamlines modify-await-chk that you would otherwise have to do separately using the fluent client. The modify and chk have to agree on the entries being programmed, and this package provides a way to batch program these entries. This is a more friendly way to write gRIBI tests.

mojiiba commented on feature/experimental/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+
+var mu sync.Mutex
+
+func NewGRIBIFluent(t testing.TB, dut *ondatra.DUTDevice, persistance, fibAck bool) *GRIBIHandler {
```
done

mojiiba commented:
@liulk We have addressed your comments and migrated the public gribi test cases to use the helpers. Please see the attached log for cisco XR. 

I will add other apis for adding route, flush, delete, etc in the next PR.

[test-cisco-log.txt](https://github.com/openconfig/featureprofiles/files/8367167/test-cisco-log.txt)

mojiiba commented:
@liulk Please let me know if you need anything else to be addressed? Please note that I do not have access to disco-idea-817 project, so I can not run the build and pass the related check.

liulk commented:
I think I will need more time to look at the TE-4.1 changes which look non-trivial. If you can split it to a separate PR, that would unblock this PR.

I'll also be doing a more thorough Go style review of the helper later today.

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+// AddNH adds an IPV4 entry
+func (g *GRIBIHandler) AddIPV4Entry(t testing.TB, nhgIndex uint64, nhgInstance, prefix, instance string, expectedResult fluent.ProgrammingResult) {
```
Rename function to AddIPv4 to be consistent with the naming pattern above? They all omitted Entry.

mojiiba replied:
done

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+// AddNH adds an IPV4 entry
+func (g *GRIBIHandler) AddIPV4Entry(t testing.TB, nhgIndex uint64, nhgInstance, prefix, instance string, expectedResult fluent.ProgrammingResult) {
```
Reorder the arguments so the mapping relationship is more clear?

```
func (g *GRIBIHandler) AddIPv4(t testing.TB, prefix string, nhgIndex uint64, instance, nhgInstance string, expectedResult fluent.ProgrammingResult)
```

mojiiba replied:
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+		t.Fatalf("learnElectionID Error: %v", err)
+	}
+	results := g.fluentC.Results(t)
+	mu.Lock()
```
Thanks for pointing this out. I was trying to use lock because of election id.  However, I noticed the gribigo client already saves the last known id, so I removed election id from the struct and the locking as well.

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+	)
+}
+
+// AddNH adds an IPV4 entry
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+	g.UpdateElectionID(t, lowElecId, highElecId)
+}
+
+// AddNHG adds a NextHOP group entry
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+
+// AddNHG adds a NextHOP group entry
+func (g *GRIBIHandler) AddNHG(t testing.TB, nhgIndex uint64, nhWeights map[uint64]uint64, instance string, expectedResult fluent.ProgrammingResult) {
+	NGH := fluent.NextHopGroupEntry().WithNetworkInstance(instance).WithID(nhgIndex)
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+	fluentClient.StartSending(context.Background(), t)
+	err := g.AwaitTimeout(context.Background(), t, timeout)
+	if err != nil {
+		t.Fatalf("can not establish the session: %v", err)
```
changed as you suggested. Thanks!

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,203 @@
+package gribi
```
added the comment

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,191 @@
+// Start function start establish a client connection with the gribi server.
+// By default the client is not the leader and for that function BecomeLeader
+// needs to be called.
+
```
Remove empty line?

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,191 @@
+
+// AddNHG adds a NextHopGroupEntry with a given index, and a map of next hop entry indices to the weights,
+// in a given network instance.
+func (g *GRIBIHandler) AddNHG(t testing.TB, nhgIndex uint64, nhWeights map[uint64]uint64, instance string,
```
Avoid line wrap?

https://github.com/golang/go/wiki/CodeReviewComments#line-length

mojiiba replied:
done

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,191 @@
+}
+
+// AddNH adds a NextHopEntry with a given index to an address within a given network instance.
+func (g *GRIBIHandler) AddNH(t testing.TB, nhIndex uint64, address, instance string,
```
Avoid line wrap?

https://github.com/golang/go/wiki/CodeReviewComments#line-length

mojiiba replied:
done

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,191 @@
+	}
+	g.fluentC.Modify().AddEntry(t, nhg)
+	if err := g.AwaitTimeout(context.Background(), t, timeout); err != nil {
+		t.Fatalf("got unexpected error from server adding a NHG, got: %v, want: nil", err)
```
Reporting error doesn't need "want: nil"

```
t.Fatalf("Error waiting to add NHG: %v", err)
```

mojiiba replied:
done

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,191 @@
+			WithIPAddress(address))
+
+	if err := g.AwaitTimeout(context.Background(), t, timeout); err != nil {
+		t.Fatalf("got unexpected error from server adding NH, got: %v, want: nil", err)
```
```
t.Fatalf("Error waiting to add NH: %v", err)
```

mojiiba replied:
done

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,191 @@
+	}
+	g.fluentC.Modify().AddEntry(t, ipv4Entry)
+	if err := g.AwaitTimeout(context.Background(), t, timeout); err != nil {
+		t.Fatalf("got unexpected error from server adding NH, got: %v, want: nil", err)
```
```
t.Fatalf("Error waiting to add IPv4: %v", err)
```

mojiiba replied:
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,188 @@
+// BecomeLeader learns the latest election id and the make the client leader by increasing the election id by one.
+func (g *GRIBIHandler) BecomeLeader(t testing.TB) {
+	t.Logf("trying to be a master with increasing the election id by one on dut: %s", g.DUT.Name())
+	lowElecId, highElecId := g.learnElectionID(t)
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,188 @@
+func (g *GRIBIHandler) BecomeLeader(t testing.TB) {
+	t.Logf("trying to be a master with increasing the election id by one on dut: %s", g.DUT.Name())
+	lowElecId, highElecId := g.learnElectionID(t)
+	if lowElecId == maxUint64 {
```
Thanks! Done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,188 @@
+	maxUint64 = 1<<64 - 1
+)
+
+// GRIBIHandler provides access to GRIBI APIs of the DUT.
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,188 @@
+	DUT         *ondatra.DUTDevice
+	FibACK      bool
+	Persistence bool
+	fluentC     *fluent.GRIBIClient
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,188 @@
+// Package gribi provides helper APIs to simplify writing gribi test cases.
```
done

mojiiba commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,188 @@
+// roles easily without a need to keep track of the server election id.
+// It also packs modify operations with the corresponding verifications to
+// prevent code duplications and increase the test code readability.
+
```
done

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+	if g.FibACK {
+		g.fluentC.Connection().WithFIBACK()
+	}
+	g.fluentC.Start(context.Background(), t)
```
```
ctx := context.Background()
```

And then reuse ctx below.

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+// Close function closes the gribi session with the dut by stopping the fluent client.
+func (g *GRIBIHandler) Close(t testing.TB) {
+	t.Helper()
+	t.Logf("closing GRIBI connection for dut: %s", g.DUT.Name())
```
Log strings are generally capitalized, unlike error strings.

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+// a dummy modify request with election id 1.
+func (g *GRIBIHandler) learnElectionID(t testing.TB) (low, high uint64) {
+	t.Helper()
+	t.Logf("learn GRIBI Election ID from dut: %s", g.DUT.Name())
```
Capitalize log string?

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+// The function fails if the requsted election id is less than the server election id.
+func (g *GRIBIHandler) UpdateElectionID(t testing.TB, lowElecId, highElecId uint64) {
+	t.Helper()
+	t.Logf("setting GRIBI Election ID for dut: %s to low=%d,high=%d", g.DUT.Name(), lowElecId, highElecId)
```
Capitalize?

liulk commented on internal/gribi/gribi.go:
```diff
@@ -0,0 +1,209 @@
+
+// BecomeLeader learns the latest election id and the make the client leader by increasing the election id by one.
+func (g *GRIBIHandler) BecomeLeader(t testing.TB) {
+	t.Logf("trying to be a master with increasing the election id by one on dut: %s", g.DUT.Name())
```
Capitalize?

liulk reviewed (changes requested):
Sorry, just noticed some more nits. Should be the last ones. Thanks for your patience!

mojiiba commented:
> Sorry, just noticed some more nits. Should be the last ones. Thanks for your patience!

Thanks a lot for the careful and thorough review!

liulk commented:
/gcbrun
